lib/gate_appl.h

```diff
@@ -35,7 +35,12 @@ template <typename Simulator, typename Gate>
 inline void ApplyGate(const Simulator& simulator, const Gate& gate,
                       typename Simulator::State& state) {
   if (gate.kind != gate::kMeasurement) {
-    simulator.ApplyGate(gate.qubits, gate.matrix.data(), state);
+    if (gate.controlled_by.size() == 0) {
+      simulator.ApplyGate(gate.qubits, gate.matrix.data(), state);
+    } else {
+      simulator.ApplyControlledGate(gate.qubits, gate.controlled_by,
+                                    gate.cmask, gate.matrix.data(), state);
+    }
```

Comment:
Would it be worthwhile for `ApplyControlledGate` to default to `ApplyGate` if there are no control qubits?

Reply:
Do you mean to use just one function, i.e. to merge `ApplyControlledGate` and `ApplyGate`?

Reply:
Either that, or conditionally calling `ApplyGate` at the start of `ApplyControlledGate`:
```
void ApplyControlledGate (...) {
  if (gate.controlled_by.size() == 0) {
    ApplyGate(...);
    return;
  }
  /* ... rest of function ... */
}
```

Reply:
Yes, that makes sense. I'll use the second option.